#[deprecated] incorrectly warns in the same crate if derives are involved

```rust

#[derive(Debug)]
#[deprecated(since = "0.1.1", note = "oh noes")]
pub struct A;

#[derive(PartialOrd)]
#[deprecated(since = "0.1.1", note = "oh noes")]
pub struct B;

fn main() {}

```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=7255596c2dc2dfdf3671bb17e98b5bc8))

Errors:

```
   Compiling playground v0.0.1 (/playground)
warning: use of deprecated item 'A': oh noes
  |
  = note: #[warn(deprecated)] on by default

warning: use of deprecated item 'B': oh noes

error[E0277]: can't compare `B` with `B`
 --> src/main.rs:6:10
  |
6 | #[derive(PartialOrd)]
  |          ^^^^^^^^^^ no implementation for `B == B`
  |
  = help: the trait `std::cmp::PartialEq` is not implemented for `B`

error: aborting due to previous error
```

An easy fix might be to `#[allow(deprecated)]` in the custom derive implementations provided by Rust. Right now one has to manually `#[allow(deprecated)]` on all deprecated structs with derives, which is a pain, and also on all their impls...

--- a/src/libsyntax_ext/deriving/generic/mod.rs
+++ b/src/libsyntax_ext/deriving/generic/mod.rs
@@ -192,7 +192,7 @@ use syntax::util::map_in_place::MapInPlace;
 use syntax::ptr::P;
 use syntax::symbol::{Symbol, kw, sym};
 use syntax::parse::ParseSess;
-use syntax_pos::{DUMMY_SP, Span};
+use syntax_pos::{Span};
 
 use ty::{LifetimeBounds, Path, Ptr, PtrTy, Self_, Ty};
 
@@ -1022,7 +1022,7 @@ impl<'a> MethodDef<'a> {
                                  // [fields of next Self arg], [etc]>
         let mut patterns = Vec::new();
         for i in 0..self_args.len() {
-            let struct_path = cx.path(DUMMY_SP, vec![type_ident]);
+            let struct_path = cx.path(trait_.span, vec![type_ident]);
             let (pat, ident_expr) = trait_.create_struct_pattern(cx,
                                                                  struct_path,
                                                                  struct_def,

--- a/src/test/ui/deprecation/derive_on_deprecated.rs
+++ b/src/test/ui/deprecation/derive_on_deprecated.rs
@@ -6,4 +6,10 @@
 #[derive(Default)]
 struct X;
 
+#[deprecated(note="Do not use this")]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default, Hash)]
+pub struct Step<I> {
+    _skip: Option<I>,
+}
+
 fn main() {}

